dev/bots/check_tests_cross_imports.dart

```diff
@@ -0,0 +1,572 @@
+      valid = false;
+      foundError(
+        _getImportError(
+          files: unknownWidgetsTestsImportingMaterial,
+          testLibraryName: _Library.widgets.name,
+          importedLibraryName: _Library.material.name,
+        ).split('\n'),
+      );
+    }
+    final Set<File> unknownWidgetsTestsImportingCupertino = _getUnknowns(
+      widgetsTestsImportingCupertino,
+      _knownCrossImports,
+    );
+    if (unknownWidgetsTestsImportingCupertino.isNotEmpty) {
+      valid = false;
+      foundError(
+        _getImportError(
+          files: unknownWidgetsTestsImportingCupertino,
+          testLibraryName: _Library.widgets.name,
+          importedLibraryName: _Library.cupertino.name,
+        ).split('\n'),
+      );
+    }
+    final Set<File> unknownCupertinoTestsImportingMaterial = _getUnknowns(
+      cupertinoTestsImportingMaterial,
+      _knownCrossImports,
+    );
+    if (unknownCupertinoTestsImportingMaterial.isNotEmpty) {
+      valid = false;
+      foundError(
+        _getImportError(
+          files: unknownCupertinoTestsImportingMaterial,
+          testLibraryName: _Library.cupertino.name,
+          importedLibraryName: _Library.material.name,
+        ).split('\n'),
+      );
+    }
+
+    // Find any known cross imports that weren't found, and are therefore fixed.
+    final Set<String> fixedWidgetsCrossImports = _differencePaths(
+      knownWidgetsCrossImports,
+      widgetsTestsImportingMaterial.union(widgetsTestsImportingCupertino),
+    );
+    if (fixedWidgetsCrossImports.isNotEmpty) {
+      valid = false;
+      foundError(_getFixedImportError(fixedWidgetsCrossImports, _Library.widgets).split('\n'));
+    }
+    final Set<String> fixedCupertinoCrossImports = _differencePaths(
+      knownCupertinoCrossImports,
+      cupertinoTestsImportingMaterial,
+    );
+    if (fixedCupertinoCrossImports.isNotEmpty) {
+      valid = false;
+      foundError(_getFixedImportError(fixedCupertinoCrossImports, _Library.cupertino).split('\n'));
+    }
+
+    return valid;
+  }
+}
+
+/// The libraries that we are concerned with cross importing.
+enum _Library {
+  widgets(directory: 'widgets', name: 'Widgets'),
+  material(directory: 'material', name: 'Material'),
+  cupertino(directory: 'cupertino', name: 'Cupertino');
+
+  const _Library({required this.directory, required this.name});
+
+  final String directory;
+  final String name;
+
+  static final RegExp cupertinoImportRegExp = RegExp(r"import 'package:flutter\/cupertino.dart'");
+  static final RegExp materialImportRegExp = RegExp(r"import 'package:flutter\/material.dart'");
+  static final RegExp widgetsImportRegExp = RegExp(r"import 'package:flutter\/widgets.dart'");
+
+  /// The RegExp that finds an import of this library.
+  RegExp get regExp {
+    return switch (this) {
+      _Library.widgets => widgetsImportRegExp,
+      _Library.material => materialImportRegExp,
+      _Library.cupertino => cupertinoImportRegExp,
+    };
+  }
```

Comment:
Should we use string contains instead of regexes here? The regexes don't seem to do any special matching.

I think it'd let us simplify this logic a bit, we could move the string patterns to a constructor argument. Something like:

```
  widgets(
    directory: 'widgets',
    name: 'Widgets',
    importPattern: "import 'package:flutter/widgets.dart'",
  ),
```